actpool/actpool.go

```diff
@@ -155,113 +183,162 @@ func (ap *actPool) AddActionEnvelopeValidators(fs ...action.SealedEnvelopeValida
-	ap.mutex.Lock()
-	defer ap.mutex.Unlock()
-
 	ctx, span := tracer.NewSpan(ctx, "actPool.Add")
 	defer span.End()
 
+	if err := checkSelpData(&act); err != nil {
+		return err
+	}
+
+	if err := ap.checkSelpWithoutState(ctx, &act); err != nil {
+		return err
+	}
+
 	// Reject action if pool space is full
-	if uint64(len(ap.allActions)) >= ap.cfg.MaxNumActsPerPool {
+
+	if uint64(ap.allActions.Count()) >= ap.cfg.MaxNumActsPerPool {
 		_actpoolMtc.WithLabelValues("overMaxNumActsPerPool").Inc()
 		return action.ErrTxPoolOverflow
 	}
-	span.AddEvent("act.IntrinsicGas")
-	intrinsicGas, err := act.IntrinsicGas()
+
+	if intrinsicGas, _ := act.IntrinsicGas(); atomic.LoadUint64(&ap.gasInPool)+intrinsicGas > ap.cfg.MaxGasLimitPerPool {
+		_actpoolMtc.WithLabelValues("overMaxGasLimitPerPool").Inc()
+		return action.ErrGasLimit
+	}
+
+	return ap.enqueue(ctx, act)
+}
+
+func checkSelpData(act *action.SealedEnvelope) error {
+	_, err := act.IntrinsicGas()
 	if err != nil {
-		_actpoolMtc.WithLabelValues("failedGetIntrinsicGas").Inc()
 		return err
 	}
-	if ap.gasInPool+intrinsicGas > ap.cfg.MaxGasLimitPerPool {
-		_actpoolMtc.WithLabelValues("overMaxGasLimitPerPool").Inc()
-		return action.ErrGasLimit
+	_, err = act.Hash()
+	if err != nil {
+		return err
 	}
-	hash, err := act.Hash()
+	_, err = act.Cost()
 	if err != nil {
 		return err
 	}
+	if act.SrcPubkey() == nil {
+		return action.ErrAddress
+	}
+	return nil
+}
+
+func (ap *actPool) checkSelpWithoutState(ctx context.Context, selp *action.SealedEnvelope) error {
+	span := tracer.SpanFromContext(ctx)
+	span.AddEvent("actPool.checkSelpWithoutState")
+	defer span.End()
+
+	hash, _ := selp.Hash()
 	// Reject action if it already exists in pool
-	if _, exist := ap.allActions[hash]; exist {
+	if _, exist := ap.allActions.Get(hash); exist {
 		_actpoolMtc.WithLabelValues("existedAction").Inc()
 		return action.ErrExistedInPool
 	}
+
 	// Reject action if the gas price is lower than the threshold
-	if act.GasPrice().Cmp(ap.cfg.MinGasPrice()) < 0 {
+	if selp.GasPrice().Cmp(ap.cfg.MinGasPrice()) < 0 {
 		_actpoolMtc.WithLabelValues("gasPriceLower").Inc()
+		actHash, _ := selp.Hash()
 		log.L().Info("action rejected due to low gas price",
-			zap.String("actionHash", hex.EncodeToString(hash[:])),
-			zap.String("gasPrice", act.GasPrice().String()))
+			zap.String("actionHash", hex.EncodeToString(actHash[:])),
+			zap.String("gasPrice", selp.GasPrice().String()))
 		return action.ErrUnderpriced
 	}
-	if err := ap.validate(ctx, act); err != nil {
-		return err
+
+	if _, ok := ap.senderBlackList[selp.SenderAddress().String()]; ok {
+		_actpoolMtc.WithLabelValues("blacklisted").Inc()
+		return errors.Wrap(action.ErrAddress, "action source address is blacklisted")
 	}
 
-	caller := act.SenderAddress()
-	if caller == nil {
-		return action.ErrAddress
+	for _, ev := range ap.actionEnvelopeValidators {
+		span.AddEvent("ev.Validate")
+		if err := ev.Validate(ctx, *selp); err != nil {
+			return err
+		}
```

Comment:
these 2 should still be in `validate()`, it is not specific to a Selp, see commented above